pkg/controller/common/rollout/workloads/deployment_scale_controller.go

```diff
@@ -0,0 +1,303 @@
+	}
+
+	if controller := metav1.GetControllerOf(s.deploy); controller != nil {
+		if controller.Kind == v1beta1.AppRolloutKind && controller.APIVersion == v1beta1.SchemeGroupVersion.String() {
+			// it's already there
+			return true, nil
+		}
+	}
+	// add the parent controller to the owner of the deployment
+	deployPatch := client.MergeFrom(s.deploy.DeepCopyObject())
+	ref := metav1.NewControllerRef(s.parentController, v1beta1.AppRolloutKindVersionKind)
+	s.deploy.SetOwnerReferences(append(s.deploy.GetOwnerReferences(), *ref))
+	s.deploy.Spec.Paused = false
+
+	// patch the deployment
+	if err := s.client.Patch(ctx, s.deploy, deployPatch, client.FieldOwner(s.parentController.GetUID())); err != nil {
+		s.recorder.Event(s.parentController, event.Warning("Failed to the start the deployment update", err))
+		s.rolloutStatus.RolloutRetry(err.Error())
+		return false, nil
+	}
+	// mark the rollout initialized
+	s.recorder.Event(s.parentController, event.Normal("Scale Initialized", "deployment is initialized"))
+	return true, nil
+}
+
+// RolloutOneBatchPods calculates the number of pods we can scale to according to the rollout spec
+func (s *DeploymentScaleController) RolloutOneBatchPods(ctx context.Context) (bool, error) {
+	err := s.fetchDeployment(ctx)
+	if err != nil {
+		s.rolloutStatus.RolloutRetry(err.Error())
+		// nolint: nilerr
+		return false, nil
+	}
+
+	deployPatch := client.MergeFrom(s.deploy.DeepCopyObject())
+	// set the replica according to the batch
+	newPodTarget := calculateNewBatchTarget(s.rolloutSpec, int(s.rolloutStatus.RolloutOriginalSize),
+		int(s.rolloutStatus.RolloutTargetSize), int(s.rolloutStatus.CurrentBatch))
+	s.deploy.Spec.Replicas = pointer.Int32Ptr(int32(newPodTarget))
+	// patch the deployment
+	if err := s.client.Patch(ctx, s.deploy, deployPatch, client.FieldOwner(s.parentController.GetUID())); err != nil {
+		s.recorder.Event(s.parentController, event.Warning("Failed to update the deployment to upgrade", err))
+		s.rolloutStatus.RolloutRetry(err.Error())
+		return false, nil
+	}
+	// record the scale
+	klog.InfoS("scale one batch", "current batch", s.rolloutStatus.CurrentBatch)
+	s.recorder.Event(s.parentController, event.Normal("Batch Rollout",
+		fmt.Sprintf("Submitted scale quest for batch %d", s.rolloutStatus.CurrentBatch)))
+	s.rolloutStatus.UpgradedReplicas = int32(newPodTarget)
+	return true, nil
+}
+
+// CheckOneBatchPods checks to see if the pods are scaled according to the rollout plan
+func (s *DeploymentScaleController) CheckOneBatchPods(ctx context.Context) (bool, error) {
+	err := s.fetchDeployment(ctx)
+	if err != nil {
+		s.rolloutStatus.RolloutRetry(err.Error())
+		// nolint:nilerr
+		return false, nil
+	}
+	newPodTarget := calculateNewBatchTarget(s.rolloutSpec, int(s.rolloutStatus.RolloutOriginalSize),
+		int(s.rolloutStatus.RolloutTargetSize), int(s.rolloutStatus.CurrentBatch))
+	// get the number of ready pod from deployment
+	// TODO: should we use the replica number when we shrink?
```

Comment:
what does it mean?

Reply:
This is copied from the clonesetController. We use the "readyReplicas" to check if the scaleup succeeded or not. it seems that we should use "replicas" to check if we scale down

Reply:
yes, check `replicas` in scale down case is better